crates/bevy_math/src/smooth.rs

```diff
@@ -0,0 +1,118 @@
+use crate::{Vec2, Vec3};
+
+/// Smooths value to a goal using a damped spring.
+pub trait SmoothDamp {
+    /// Smooths value to a goal using a damped spring.
+    ///
+    /// `smooth_time` is the expected time to reach the target when at maximum velocity.
+    ///
+    /// Returns smoothed value and new velocity.
+    fn smooth_damp(
+        from: Self,
+        to: Self,
+        velocity: Self,
+        smooth_time: f32,
+        delta_time: f32,
+    ) -> (Self, Self)
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_smooth_damp {
+    ($t:ty, $f:ty) => {
+        impl SmoothDamp for $t {
+            fn smooth_damp(
+                from: $t,
+                to: $t,
+                velocity: $t,
+                smooth_time: f32,
+                delta_time: f32,
+            ) -> ($t, $t) {
+                let smooth_time = <$f>::max(smooth_time as $f, 0.0001); // ensure smooth_time is positive and non-zero
+                let delta_time = delta_time as $f;
+
+                // from game programming gems 4, chapter 1.10
+                let omega = 2.0 / smooth_time;
+                let x = omega * delta_time;
+
+                let exp = 1.0 / (1.0 + x + 0.48 * x * x + 0.235 * x * x * x);
+                // let exp = 1.0 / (1.0 + x * (1.0 + x * (0.48 + 0.235 * x))); // TODO: profile me, both in debug & release
+
+                let change = from - to;
+                let temp = (velocity + omega * change) * delta_time;
+
+                (
+                    to + (change + temp) * exp,      // position
+                    (velocity - omega * temp) * exp, // velocity
+                )
+            }
+        }
+    };
+}
+
+impl_smooth_damp! {f32, f32}
+impl_smooth_damp! {f64, f64}
+impl_smooth_damp! {Vec2, f32}
+impl_smooth_damp! {Vec3, f32}
+
+/// Smooths value to a goal using a damped spring limited by a maximum speed.
+pub trait SmoothDampMax {
+    /// Smooths value to a goal using a damped spring limited by a maximum speed.
+    ///
+    /// `smooth_time` is the expected time to reach the target when at maximum velocity.
+    ///
+    /// Returns smoothed value and new velocity.
+    fn smooth_damp_max(
+        from: Self,
+        to: Self,
+        velocity: Self,
+        max_speed: f32,
+        smooth_time: f32,
+        delta_time: f32,
+    ) -> (Self, Self)
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_smooth_damp_max {
+    ($t:ty, $f:ty, $clamp:expr) => {
+        impl SmoothDampMax for $t {
+            fn smooth_damp_max(
+                from: $t,
+                to: $t,
+                velocity: $t,
+                max_speed: f32,
+                smooth_time: f32,
+                delta_time: f32,
+            ) -> ($t, $t) {
+                let max_speed = <$f>::max(max_speed as $f, 0.0001); // ensure max speed is positive and non-zero
```

Comment:
I'm curious if this is the best behavior here.
Currently I see 3 options.
1) As is where we take the max between the value and some arbitrary value `0.0001`
2) `panic!` is the value is <= 0. (Or `assert!`)
3) Have this return a `Result<T, E>` in the case where the value is <= 0.

Reply:
Option 1 seems the best to me as the cost of this max is negligible and it ensures the function plays well even when the parameters are generated.

Reply:
Personally, I like option 2, since IMO, if a user is passing in a value `<= 0`, then they are using this function incorrectly and we should inform them about this and not just silently succeed. 

For example, even in the `f64::clamp` [function](https://doc.rust-lang.org/std/primitive.f32.html#method.clamp), the behavior is to `panic!` if the `min > max`. This is help ensure people use this function correctly, as correctness tends to be the most important for Rust code.

Reply:
Fair enough! Asserts it is